### PR TITLE
LTD-306: Flagging and routing rules changes

### PR DESCRIPTION
**Pull request by saruniitr**

### CODE REVIEW

File: caseworker/cases/services.py [hunk 1]
```diff
@@ -270,7 +270,7 @@ def get_flags_for_team_of_level(request, level, team_id, include_system_flags=Fa
     :param include_system_flags: used to indicate adding system flags to list of team flags returned
     :return:
     """
-    data = client.get(request, "/flags/" + convert_parameters_to_query_params(locals()) + "&disable_pagination=True")
+    data = client.get(request, "/flags/" + convert_parameters_to_query_params(locals()) + "&disable_pagination=True&include_flagging_rules=True")
```
Comment (saruniitr) [inline caseworker/cases/services.py:273]: When adding flags to routing rules, design requires displaying their flagging rules also hence they are pulled conditionally based on this parameter.

File: caseworker/routing_rules/forms.py [hunk 2]
```diff
@@ -73,20 +90,80 @@ def select_case_type(request):
...
+                    Option(key="contain_selected_flags", value="Contain selected flags"),
+                    Option(key="doesnot_contain_selected_flags", value="Do not contain selected flags"),
+                ],
+            ),
+        ],
+        buttons=[
+            Button("Save and continue", "submit", id="save_and_continue"),
+            Button("Add another condition", "", ButtonStyle.SECONDARY, link="#"),
         ],
+        javascript_imports={"/javascripts/routing-rules-flags.js"},
```
Comment (saruniitr) [inline caseworker/routing_rules/forms.py:166]: There is no standard component available to implement the design required for Routing rules flags section, that is implemented in this script. The above function `get_flag_details_html()` is also used for that purpose to generate the details in side menu

File: caseworker/routing_rules/forms.py [hunk 3]
```diff
@@ -73,20 +90,80 @@ def select_case_type(request):
...
                 name="flags[]",
                 options=[
-                    Option(flag["id"], flag["name"])
+                    Option(flag["id"], flag["name"], data_attribute=get_flag_details_html(flag))
                     for flag in get_flags_for_team_of_level(
                         request, level="", team_id=team_id, include_system_flags=True
                     )[0]
                 ],
-            )
+                import_custom_js=["/javascripts/filter-checkbox-list-flags.js"],
```
Comment (saruniitr) [inline caseworker/routing_rules/forms.py:150]: Existing filter-checkbox-list displays title in a different way than required in the design hence it is created to customize that.

File: caseworker/routing_rules/views.py [hunk 4]
```diff
@@ -79,11 +82,14 @@ class CreateRoutingRule(LoginRequiredMixin, MultiFormView):
     def init(self, request, **kwargs):
         select_team = has_permission(request, Permission.MANAGE_ALL_ROUTING_RULES)
         team_id = request.POST.get("team", get_gov_user(request)[0]["user"]["team"]["id"])
+        if not team_id:
+            team_id = get_gov_user(request)[0]["user"]["team"]["id"]
+        additional_rules = request.POST.getlist("additional_rules[]", ())
+        print(f"==> Additional rules: {additional_rules}")
```
Comment (saruniitr) [inline caseworker/routing_rules/views.py:88]: will remove

File: caseworker/routing_rules/forms.py [hunk 5]
```diff
@@ -73,20 +90,80 @@ def select_case_type(request):
...
+        </details>
+    """
+
+
+def select_flags(request, team_id, flags_to_include, flags_to_exclude):
     return Form(
         title=Forms.FLAGS,
         questions=[
+            HiddenField(name="flags_to_include", value=",".join(flags_to_include)),
+            HiddenField(name="flags_to_exclude", value=",".join(flags_to_exclude)),
```
Comment (saruniitr) [inline caseworker/routing_rules/forms.py:139]: The flags selected by the user are added a comma separated values to these fields.

File: core/assets/javascripts/routing-rules-flags.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
+(function () {
+    var flagsConditions = [];
```
Comment (saruniitr) [inline core/assets/javascripts/routing-rules-flags.js:2]: this is the script to implement the routing rules flags section.

File: core/assets/javascripts/routing-rules-flags.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+    function renderTable(conditions) {
```
Comment (saruniitr) [inline core/assets/javascripts/routing-rules-flags.js:4]: displays the selected flags along with their conditions in a tabular form and option to remove them.

File: core/assets/javascripts/routing-rules-flags.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        renderTable(flagsConditions);
+    }
+
+    function getFlagIds(condition, flags) {
+        var filtered = flags.filter(function (flag) { return flag.condition === condition; })
+        var allIds = filtered.map(function (item) { return item.ids; })
+        return allIds;
+    }
+
+    function addConditionWithSelectedFlags() {
```
Comment (saruniitr) [inline core/assets/javascripts/routing-rules-flags.js:71]: collects the selected flags and updates a list. User can add multiple conditions with different flags.

**Review by richtier — APPROVED** [review]

**Pull request merged**